Weighted uniform random tipselection + COO heaviest branch tipselection

muXxer opened the pull request:
# Description

This PR adds the Weighted URTS and the COO heaviest branch tipselect for chrysalis step 1.

# ToDo
- [x] Propagate Root txs snapshot indexes in the future cone
- [x] Check the cone for invalid bundles and cache that info
- [x] Question: Do lazy tips remove other tips from the pool?
- [x] Run the solidifier in correct order
- [ ] Solve COO tipselect todos
- [ ] Write test cases

Wollac committed:
use heaviest branch for milestones

muXxer committed:
Add first URTS draft

alexsporn commented on config.json:
```diff
@@ -99,8 +99,11 @@
   },
   "tipsel": {
-    "belowMaxDepthTransactionLimit": 20000,
-    "maxDepth": 5
```
why rename maxDepth to belowMaxDepth?

muXxer replied:
maxDepth was another thing back then... it was used for maximum starting depth of the random walk.
This belowMaxDepth is the treshold for lazy tips.

muXxer committed:
Remove checkConsistency from permitted API calls

muXxer committed:
Solidify txs in order from oldest to latest

muXxer committed:
Add bundle MetadataInvalidPastCone

muXxer committed:
Set root snapshot indexes at confirmation

muXxer committed:
Add check for invalid past cone

muXxer committed:
Skip check for confirmed tx in getTransactionRootSnapshotIndexes

muXxer committed:
Adapt TipSelStats for URTS

muXxer committed:
Check tips again when they get picked during tipselect

muXxer committed:
Add helper function TraverseApprovers to walk the future cone

muXxer committed:
Add TailsReferenced set to whiteflag

muXxer committed:
Add MilestoneConfirmed event to tangle plugin

muXxer committed:
Refactor URTS events

muXxer committed:
Update transaction root snapshot index of the future cone

muXxer committed:
Fix the order of the walk in UpdateTransactionRootSnapshotIndexes

muXxer committed:
Update root transaction snapshot indexes with correct LSMI

muXxer committed:
Use hornet.Hash instead of trinary.Hash in coordinator pkg

muXxer committed:
Exclude conflicting Txs in the visualizer

muXxer committed:
Merge branch 'chrysalis' of github.com:gohornet/hornet into feat/urts

muXxer committed:
Pass func pointer for onConfirmedMilestone to ConfirmMilestone

muXxer committed:
Remove unused func parameter in ComputeMerkleTreeRootHash

alexsporn commented on pkg/model/hornet/transaction_metadata.go:
```diff
@@ -124,6 +127,23 @@ func (m *TransactionMetadata) SetConflicting(conflicting bool) {
+	m.SetModified(true)
+}
+
+func (m *TransactionMetadata) GetRootSnapshotIndexes() (milestone.Index, milestone.Index, milestone.Index) {
```
maybe give the return values a label?
`(yrtsi milestone.Index, ortsi milestone.Index, rtsci milestone.Index)`

alexsporn commented on pkg/model/mselection/heaviest.go:
```diff
@@ -0,0 +1,193 @@
+	"github.com/gohornet/hornet/pkg/model/tangle"
+	"github.com/gohornet/hornet/pkg/utils"
+	"github.com/iotaledger/iota.go/trinary"
+	"github.com/willf/bitset"
```
order the imports

alexsporn commented on pkg/model/mselection/heaviest.go:
```diff
@@ -0,0 +1,193 @@
+	ErrWrongReference = errors.New("reference does not match root")
+)
+
+const capIncreaseFactor = 1.15
```
Comment what this controls

muXxer committed:
Resolve some review comments

muXxer committed:
Add parameter names to GetRootSnapshotIndexes

alexsporn commented on plugins/tangle/solidifier.go:
```diff
@@ -383,6 +386,8 @@ func solidifyMilestone(newMilestoneIndex milestone.Index, force bool) {
 	conf, err := whiteflag.ConfirmMilestone(cachedMsToSolidify.Retain(), func(tx *tangle.CachedTransaction, index milestone.Index, confTime int64) {
 		Events.TransactionConfirmed.Trigger(tx, index, confTime)
+	}, func(confirmation *whiteflag.Confirmation) {
+		Events.MilestoneConfirmed.Trigger(confirmation)
```
Move 
```
tangle.SetSolidMilestoneIndex(conf.Index)
Events.SolidMilestoneChanged.Trigger(cachedMsToSolidify) // bundle pass +1
```
to inside this callback

muXxer committed:
Move the SolidMilestoneChanged event before MilestoneConfirmed

muXxer committed:
Raise the critical error

alexsporn commented on pkg/tipselect/urts.go:
```diff
@@ -0,0 +1,485 @@
+}
+
+const (
+	// ScoreLazy is a lazy zip and should not be selected.
```
tip not zip :P

alexsporn commented on pkg/tipselect/urts.go:
```diff
@@ -0,0 +1,485 @@
+	// ScoreLazy is a lazy zip and should not be selected.
+	ScoreLazy Score = iota
+	// ScoreSemiLazy is a somewhat lazy tip.
+	ScoreSemiLazy
```
do you want to call is `ScoreSomewhatLazy`? It sounds weird, but uses the terminology from the RFC

alexsporn replied:
OK, the RFC mixes up definitions and call it semi lazy later

alexsporn commented on pkg/tipselect/urts.go:
```diff
@@ -0,0 +1,485 @@
+	if tip, exists := ts.tipsMap[string(tailTxHash)]; exists {
+		ts.scoreSum -= int(tip.Score)
+		delete(ts.tipsMap, string(tailTxHash))
+		ts.Events.TipRemoved.Trigger(tip)
```
sure we should trigger an event here while locked?

muXxer replied:
Do you have an idea how to do that, without making everything unreadable and complex? I unfortunately don't have one :(

alexsporn commented on pkg/tipselect/urts.go:
```diff
@@ -0,0 +1,485 @@
+		// if randScore reaches zero or below, we return the given tip
+		if randScore <= 0 {
+			// check the "own" score of the tip again to avoid old tips
+			if score := ts.calculateScore(tip.Hash, false, lsmi); score == ScoreLazy {
```
this is not done in the RFC

alexsporn replied:
OK, I see you use it to clean up the tip pool

muXxer committed:
Fix coordinator network bootstrapping

alexsporn commented on pkg/tipselect/urts.go:
```diff
@@ -0,0 +1,485 @@
+	}
+
+	// if all available approvees' OTRSI violates the LSMI delta in relation to C2 the tip is lazy too
+	if len(approveeHashes) == approveesLazy {
```
I like you using len here instead of hardcoded 2 as in the RFC, because this way you handle trunk == branch 👍

muXxer replied:
Exactly, RFC is not optimal in that case :)

muXxer committed:
Ignore solid entry points approvees in calculate score

muXxer committed:
Remove worker for gossip solidifier

muXxer committed:
Fix branchVisited if trunkVisited and trunkAndBranchAreEqual

muXxer committed:
Fix infinite loop in getTransactionRootSnapshotIndexes

muXxer committed:
Fix leak in solidQueueCheck

muXxer committed:
Add conflicting txs to the visualizer

muXxer committed:
Fix visualizer confirmation

muXxer committed:
Add badge for tip color

muXxer committed:
Add skipStartTx to FindAllTails

muXxer committed:
Add shared metric for tips

muXxer committed:
Revert "Skip check for confirmed tx in getTransactionRootSnapshotIndexes"

This reverts commit c645ede445f318df7d3e52d9bc31859e0960e913.

muXxer committed:
Only check OTRSI to LSMI delta of approvees

muXxer committed:
Remove depth from spammer config

muXxer committed:
Change tipsel default configs

muXxer committed:
Fix concurent map write and iteration

muXxer committed:
Fix typo

alexsporn commented on pkg/whiteflag/white_flag.go:
```diff
@@ -248,11 +273,13 @@ func ProcessStack(stack *list.List, wfConf *Confirmation, visited map[string]str
 		}
 
 		if conflicting {
+			wfConf.TailsReferenced = append(wfConf.TailsReferenced, currentTx.GetTxHash())
```
move this before the if

alexsporn commented on pkg/whiteflag/white_flag.go:
```diff
@@ -248,11 +273,13 @@ func ProcessStack(stack *list.List, wfConf *Confirmation, visited map[string]str
 		}
 
 		// mark the given tail to be part of milestone ledger changing tail inclusion set
+		wfConf.TailsReferenced = append(wfConf.TailsReferenced, currentTx.GetTxHash())
```
if moved before the if, remove 2nd stmt

muXxer committed:
Code cleanup

alexsporn committed:
- Switched HPS to use solid bundles instead of solid tx events

alexsporn committed:
- Fixed comments

alexsporn committed:
- Renamed solid bundle closure and moved invalid bundle check to the coo plugin and out of mselection

muXxer committed:
Remove checkApprovees special case

muXxer committed:
Remove root transaction from HPS

This commit removes the root transaction from the heaviest pair selector and changes to logic to return the heaviest branch (tip with the most transactions approved).

muXxer committed:
Fix RandomInsecure range

muXxer committed:
Disable HPS tests

muXxer committed:
Change comnet snapshot downloadURL
